### PR TITLE
[NoQA] Web job for the testBuild workflow

### PR DESCRIPTION
**Pull request by staszekscp**
### Details

The PR adds a job to build the web testing app in the `testBuild` workflow

### Fixed Issues

$ GH_LINK   
PROPOSAL: GH_LINK_ISSUE(COMMENT)

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Review by luacmartins — DISMISSED** [review]
LGTM but I'd definitely like more eyes since these workflows aren't my strength

**Comment by AndrewGable** [comment]
Please do not merge - I need to confirm with the infra team this is how we want to set it up.

**Comment by PauloGasparSv** [comment]
Is this P.R. ready for review? I saw some changes earlier today and no GH issue linked so I was waiting for an update to start reviewing.
Thx for adding people here @luacmartins!

**Comment by AndrewGable** [comment]
@coleaeason said he would review the infra side of things, thank you @coleaeason ❤️ !!

File: .github/scripts/createCloudfrontDistribution.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+if [[ $(aws cloudfront list-distributions --query "DistributionList.Items[0].Id") != null ]] && [[ $DISTRIBUTION_ID ]] ; then 
+    echo "Distribution for PR #$1 already exists! Invalidating cache..."
+    aws cloudfront create-invalidation --distribution-id $DISTRIBUTION_ID --paths '/'
+    exit 0;
+else
+    echo "A new distribution for PR #$1 is being created"
+    echo $(aws cloudfront create-distribution --origin-domain-name ad-hoc-expensify-cash.s3.us-east-1.amazonaws.com --default-root-object index.html) >> cloudfront.config.json
+
+    CONFIG=$(cat "./cloudfront.config.json")
```
Comment (coleaeason) [inline .github/scripts/createCloudfrontDistribution.sh:14]: if you do my above suggestion you can delete this line

File: .github/scripts/createCloudfrontDistribution.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    DISTRIBUTION_ID=$(echo $CONFIG | jq -r .Distribution.Id)
```
Comment (coleaeason) [inline .github/scripts/createCloudfrontDistribution.sh:15]: nab, you can do:

```bash
jq -r .Distribution.Id <<< $CONFIG
```

File: .github/scripts/createCloudfrontDistribution.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    echo "Distribution for PR #$1 already exists! Invalidating cache..."
+    aws cloudfront create-invalidation --distribution-id $DISTRIBUTION_ID --paths '/'
+    exit 0;
+else
+    echo "A new distribution for PR #$1 is being created"
+    echo $(aws cloudfront create-distribution --origin-domain-name ad-hoc-expensify-cash.s3.us-east-1.amazonaws.com --default-root-object index.html) >> cloudfront.config.json
+
+    CONFIG=$(cat "./cloudfront.config.json")
+    DISTRIBUTION_ID=$(echo $CONFIG | jq -r .Distribution.Id) 
+    ETAG=$(echo $CONFIG | jq -r .ETag)
```
Comment (coleaeason) [inline .github/scripts/createCloudfrontDistribution.sh:16]: same here, use a herestring

File: .github/scripts/createCloudfrontDistribution.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    exit 0;
+else
+    echo "A new distribution for PR #$1 is being created"
+    echo $(aws cloudfront create-distribution --origin-domain-name ad-hoc-expensify-cash.s3.us-east-1.amazonaws.com --default-root-object index.html) >> cloudfront.config.json
+
+    CONFIG=$(cat "./cloudfront.config.json")
+    DISTRIBUTION_ID=$(echo $CONFIG | jq -r .Distribution.Id) 
+    ETAG=$(echo $CONFIG | jq -r .ETag) 
+
+    echo $CONFIG | jq -r .Distribution.DistributionConfig >> dist.config.json
```
Comment (coleaeason) [inline .github/scripts/createCloudfrontDistribution.sh:18]: why are we saving a json file here? Can't we just use a bash variable?

File: .github/scripts/createCloudfrontDistribution.sh [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+    tmp=$(mktemp /tmp/tmp.XXXXXXX)
+    NEW_ORIGIN_PATH=$(echo "/web/$1")
+    jq --arg originPath "$NEW_ORIGIN_PATH" '(.Origins.Items[] | select(.OriginPath == "")).OriginPath |= $originPath' dist.config.json  > "$tmp" && mv "$tmp"  dist.config.json
```
Comment (coleaeason) [inline .github/scripts/createCloudfrontDistribution.sh:22]: I'm not sure I understand why we need this tmp file?

**Review by coleaeason — CHANGES_REQUESTED** [review]

**Comment by roryabraham** [comment]
NAB but I put up a PR [here](https://github.com/Expensify/App/pull/13387) to lint bash scripts in this repo with ShellCheck, and it might be beneficial to make sure that the script you created here is passing ShellCheck using:

```bash
npx shellcheck -e SC1091 .github/scripts/createCloudfrontDistribution.sh
```

File: .github/scripts/createCloudfrontDistribution.sh [hunk 6]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+set -e
+
+DISTRIBUTION_ID=$(echo aws cloudfront list-distributions --query "DistributionList.Items[?Origins.Items[?OriginPath=='/web/$1']].Id" --output text)
```
[the remark this answers is not in the document]
Comment (staszekscp) [reply .github/scripts/createCloudfrontDistribution.sh]: Removed, thank you, I missed this one!

File: .github/scripts/createCloudfrontDistribution.sh [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+if [[ $(aws cloudfront list-distributions --query "DistributionList.Items[0].Id") != null ]] && [[ $DISTRIBUTION_ID ]] ; then
```
[the remark this answers is not in the document]
Comment (staszekscp) [reply .github/scripts/createCloudfrontDistribution.sh]: Actually a string "null" is returned, that's why I was checking it. I've just updated the script by adding the `"`

File: .github/scripts/createCloudfrontDistribution.sh (hunk 4 above)
Comment (staszekscp) [reply .github/scripts/createCloudfrontDistribution.sh]: `aws cloudfront update-distribution` command requires sending a file with the config, that's why I decided to create it

File: .github/scripts/createCloudfrontDistribution.sh (hunk 5 above)
Comment (staszekscp) [reply .github/scripts/createCloudfrontDistribution.sh:21]: As far as I know `jq` can't edit files, that's why I create a temporary file in order to make a change

**Comment by staszekscp** [comment]
@coleaeason thank you for the review! I've added a couple of changes to address your comments. @roryabraham I have also run linter agains the script and it passes the check. :)

**Review by luacmartins — DISMISSED** [review]

**Comment by AndrewGable** [comment]
@coleaeason - Will this work with the current token? I thought the AWS token was scoped to only S3?

**Comment by AndrewGable** [comment]
Apologies, we believe we are going to go a different direction and use Cloudflare workers instead of Cloudfront as mentioned here: https://expensify.slack.com/archives/C04878MDF34/p1671061957278119

@coleaeason will circle back soon with how this is all going to tie together, thank you for your patience!

File: .github/workflows/testBuild.yml [hunk 8]
```diff
@@ -189,10 +189,42 @@ jobs:
...
+      - name: Build docs
+        run: npm run storybook-build
+        continue-on-error: true
+
+      - name: Deploy to S3 for internal testing
+        run: aws s3 cp --recursive --acl public-read "$GITHUB_WORKSPACE"/dist s3://ad-hoc-expensify-cash/web/"$PULL_REQUEST_NUMBER"
+
+      - name: Create CloudFront distribution for the PR if needed
+        run: ./.github/scripts/createCloudfrontDistribution.sh "$PULL_REQUEST_NUMBER"
+
```
Comment (coleaeason) [inline .github/workflows/testBuild.yml:223]: As Andrew mentioned, we are going to remove Cloudfront form this setup, so you can remove this step.

File: .github/scripts/createCloudfrontDistribution.sh [hunk 9]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Comment (coleaeason) [inline .github/scripts/createCloudfrontDistribution.sh:1]: This can also be removed now

File: .github/workflows/testBuild.yml [hunk 10]
```diff
@@ -217,6 +249,12 @@ jobs:
           content_ios="${content_ios//$'\r'/'%0D'}"
           echo "ios_paths=$content_ios" >> "$GITHUB_OUTPUT"
       
+      - name: Get CloudFront URL
+        id: get_cloudfront_url
+        run: |
+          cloudfront_url=$(aws cloudfront list-distributions --query "DistributionList.Items[?Origins.Items[?OriginPath=='/web/$PULL_REQUEST_NUMBER']].DomainName" --output text)
+          echo "cloudfront_url=$cloudfront_url" >> "$GITHUB_OUTPUT"
+
```
Comment (coleaeason) [inline .github/workflows/testBuild.yml:257]: This can also be removed.

File: .github/workflows/testBuild.yml [hunk 11]
```diff
@@ -232,9 +270,9 @@ jobs:
...
-          | ${{fromJson(steps.set_var.outputs.android_paths).html_path}}  | ${{fromJson(steps.set_var.outputs.ios_paths).html_path}}  | https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/$PULL_REQUEST_NUMBER/NewExpensify.dmg |
-          | ![Android](https://api.qrserver.com/v1/create-qr-code/?size=200x200&data=${{fromJson(steps.set_var.outputs.android_paths).html_path}}) | ![iOS](https://api.qrserver.com/v1/create-qr-code/?size=200x200&data=${{fromJson(steps.set_var.outputs.ios_paths).html_path}}) | ![desktop](https://api.qrserver.com/v1/create-qr-code/?size=200x200&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/$PULL_REQUEST_NUMBER/NewExpensify.dmg) |"
+          | android :robot:  | iOS :apple: | desktop :computer: | web :spider_web: |
+          | ------------- | ------------- | ------------- | ------------- | 
+          | ${{fromJson(steps.set_var.outputs.android_paths).html_path}}  | ${{fromJson(steps.set_var.outputs.ios_paths).html_path}}  | https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/$PULL_REQUEST_NUMBER/NewExpensify.dmg | https://${{steps.get_cloudfront_url.outputs.cloudfront_url}} |
```
Comment (coleaeason) [inline .github/workflows/testBuild.yml:275]: The URL now is:

```
https://<PR_NUMBER>.pr-testing.expensify.com
```

**Review by coleaeason — CHANGES_REQUESTED** [review]

**Comment by staszekscp** [comment]
@coleaeason Thanks for the info, I've just applied the changes!

**Review by coleaeason — APPROVED** [review]
LGTM -- Not sure why checks are failing. @AndrewGable ?

**Comment by luacmartins** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Comment by luacmartins** [comment]
@staszekscp we've updated the author checklist. Could you please update your checklist accordingly?

**Review by luacmartins — APPROVED** [review]

**Comment by staszekscp** [comment]
@luacmartins Done!

**Review by AndrewGable — APPROVED** [review]

**Comment by AndrewGable** [comment]
Testing here: https://github.com/Expensify/App/actions/runs/3716737239

**Comment by staszekscp** [comment]
@AndrewGable @luacmartins @coleaeason The link works, that is cool! Great job guys! https://13354.pr-testing.expensify.com/

**Comment by luacmartins** [comment]
Nice! Awesome job @staszekscp!

**Pull request merged**